generators.py

```diff
@@ -187,6 +200,83 @@ def generate_pendulum_intensity_sequence(args):
 
     return sequence, thetas, velocities
 
+def generate_bouncing_ball_drop_sequence(args):
+    sequence = []
+    height = np.random.uniform(args.r, args.img_size - args.r) # put ball in image
+    velocity = 0
+
+    heights = []
+    velocities = []
+
+    for _ in range(args.seq_len):
+        heights.append(height)
+        velocities.append(velocity)
+
+        frame = np.zeros((args.img_size, args.img_size, 3))
+
+        clamped_height = np.clip(int(height), args.r, args.img_size - args.r)
+
+        rr, cc = disk((args.img_size - int(clamped_height), args.img_size // 2), args.r)
+        frame[rr, cc, :] = (255, 0, 0)
+        frame = frame.astype(np.uint8)
+
+        sequence.append(frame)
+    
+        for _ in range(args.ode_steps):
+            height += velocity * (args.dt / args.ode_steps)
+            velocity -= args.g * (args.dt / args.ode_steps)
+
+            if (height <= args.r): # check for bounce
+                height = args.r
+                velocity = -velocity * args.elasticity # reverse velocity (changing direction) and simulate energy loss of elasticity
+
+    return sequence, heights, velocities
+
+def generate_ball_throw_sequence(args):
+    sequence = []
+    initial_velocity = np.random.uniform(0, args.max_velocity)
+    theta = args.theta
+
+    velocity_x = initial_velocity * np.cos(theta)
+    velocity_y = initial_velocity * np.sin(theta)
+
+    position_x = args.r
+    position_y = args.img_size // 2
+
+    positions_x = []
+    positions_y = []
+    velocities_x = []
+    velocities_y = []
+
+    for _ in range(args.seq_len):
+        positions_x.append(position_x)
+        positions_y.append(position_y)
+        velocities_x.append(velocity_x)
+        velocities_y.append(velocity_y)
+
+        frame = np.zeros((args.img_size, args.img_size, 3))
+
+        clamped_position_y = np.clip(int(position_y), args.r, args.img_size - args.r)
+        clamped_position_x = np.clip(int(position_x), args.r, args.img_size - args.r)
+
+        rr, cc = disk((args.img_size - int(clamped_position_y), int(clamped_position_x)), args.r)
+        frame[rr, cc, :] = (255, 0, 0)
+        frame = frame.astype(np.uint8)
+
+        sequence.append(frame)
+    
+        for _ in range(args.ode_steps):
+            position_x += velocity_x * (args.dt / args.ode_steps)
+            position_y += velocity_y * (args.dt / args.ode_steps)
+
+            if (position_y <= args.r):
+                velocity_x = 0
+                velocity_y = 0
+                position_y = args.r
+            else:
+                velocity_y -= args.g * (args.dt / args.ode_steps)
+    return sequence, positions_x, positions_y
```

Comment:
We should always return position and velocity. Consider making position (and velocity) an array of tuples (or a 2D array; an array of tuples might not work with numpy).

Reply:
Ah good idea, will do!

Reply:
I now return the zips of the two. So a list of tuples.